### PR TITLE
Patterns: check that pattern is synced before replacing blocks with synced pattern on creation

### LINKED ISSUE #54803: Patterns: ignoring sync option and causing strange effects in syncing status
### Description

Using GB 16.7 RC 2 and WordPress 6.3.1 OR 6.4-alpha-56670, I'm unable to create an unsynced pattern. When I deselect the sync option, a synced pattern is still created. cc @aaronrobertshaw. 

### Step-by-step reproduction instructions

1. Open the Site editor.
2. Head to Pages.
3. Open a page.
4. Create some blocks.
5. Select those blocks and use the three dot menu to create a pattern in the Site Editor. 
6. Add a category. 
7. Turn off the sync option.
8. Notice the snackbar notif says unsynced pattern created but notice a synced pattern shows up both in the inserter and on the page itself. 

### Screenshots, screen recording, code snippet

Here's a short vid showing the problem at its simplest:

https://github.com/WordPress/gutenberg/assets/26996883/e31ebeaf-bc2d-41f4-8d77-6e69996182c4

Here's me talking through the bugginess as I'm not quite sure how best to explain it otherwise! Basically the "syncness" seems to not stick and randomly ignore sync status.

https://github.com/WordPress/gutenberg/assets/26996883/c8667871-ab22-4936-9959-a426df5d9c28

### Environment info

- 6.3.1 and 6.4-alpha-56670
- GB 16.7 RC 2

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by glendaviesnz**
## What?
Adds a check to the pattern creation code to make sure the new pattern is synced before replacing the existing blocks in the editor canvas.

## Why?
After a change in the [pattern creation modal to account for the import/export of patterns via JSON](https://github.com/WordPress/gutenberg/pull/54337/files#diff-57597c4236082d67c5cbe7c070695e6803d0961b99d8d84d5ff37d55c38a1983R99) new unsynced patterns are initially being inserted into the post editor canvas as synced even though they entity is successfully saved as unsynced.

Fixes: https://github.com/WordPress/gutenberg/issues/54803

## How?
Adds a check in the create pattern button success method to only replace the blocks in the editor canvas if the new pattern is not unsynced.

## Testing Instructions

- In the post editor add a block, and then in the block toolbar overflow menu select `Create pattern`
- Set the new pattern sync status to unsynced and make sure that the new pattern is added but the existing block in the canvas is not converted to a synced pattern
- Repeat but this time created a synced pattern and make sure the block on the editor canvas is replaced with a synced pattern
- In the site editor create a synced and an unsynced pattern and make sure that works as expected
- In the site editor export and import both synced and unsynced patterns and make sure that works as expected

## Screenshots or screencast 

Before:

https://github.com/WordPress/gutenberg/assets/3629020/46742a7b-8673-44ca-a4bd-6c9d09918eca

After:

https://github.com/WordPress/gutenberg/assets/3629020/e6b0bb07-0353-4c6e-a222-151d898cf1b9

### CODE REVIEW

**Review by ramonjd — APPROVED** [review]
- I can create unsynced and synced patterns in the Post and Site editors, both are available in the Block inserter menu
- Exporting and importing unsynced and synced patterns works in the site editor

LGTM

**Comment by glendaviesnz** [comment]
Looks like there is a [fix for the failing unit tests here](https://github.com/WordPress/gutenberg/pull/54806) - will rebase once that is merged.

**Review by aaronrobertshaw — APPROVED** [review]
Thanks for the rapid turnaround in fixing this @glendaviesnz 👍 

✅ Post Editor: Creation of unsynced and synced patterns works as expected
✅ Site Editor: Creation of unsynced and synced patterns continues to function
✅ Import/Export: Works as before also

**Comment by glendaviesnz** [comment]
This is a reasonably critical fix that should be in the 6.4 Beta, and we are reasonably confident that it is not the cause of the current test failures, which are discussed [here](https://github.com/WordPress/gutenberg/pull/54806) and [here](https://wordpress.slack.com/archives/C02QB2JS7/p1695664602044649), so going to go ahead and merge so there is no hold up with getting this into 17.6 and beta. It is a reasonably small change so should be easy to follow up if any issues we have overlooked.

**Comment by mikachan** [comment]
I just cherry-picked this PR to the release/16.7 branch to get it included in the next release: fb8ffc9b9d

**Pull request merged**
